dev/usb/class/cdcserial/cdcserial.c

```diff
@@ -0,0 +1,331 @@
+    0x00,       /* iInterface: */
+
+    /*Endpoint OUT Descriptor*/
+    0x07,      /* bLength: Endpoint Descriptor size */
+    ENDPOINT,  /* bDescriptorType: Endpoint */
+    0x01,      /* bEndpointAddress */
+    0x02,      /* bmAttributes: Bulk */
+    0x40,      /* wMaxPacketSize: */
+    0x00,
+    0x00,      /* bInterval: ignore for Bulk transfer */
+
+    /*Endpoint IN Descriptor*/
+    0x07,      /* bLength: Endpoint Descriptor size */
+    ENDPOINT,  /* bDescriptorType: Endpoint */
+    0x81,      /* bEndpointAddress */
+    0x02,      /* bmAttributes: Bulk */
+    0x40,      /* wMaxPacketSize: */
+    0x00,
+    0x00       /* bInterval */
+};
+
+static const uchar langid[] = { 0x04, 0x03, 0x09, 0x04 };
+
+usb_config config = {
+    .lowspeed = {
+        .device = USB_DESC_STATIC(dev_descr),
+        .device_qual = USB_DESC_STATIC(devqual_descr),
+        .config = USB_DESC_STATIC(cfg_descr),
+    },
+    .highspeed = {
+        .device = USB_DESC_STATIC(dev_descr),
+        .device_qual = USB_DESC_STATIC(devqual_descr),
+        .config = USB_DESC_STATIC(cfg_descr),
+    },
+
+    .langid = USB_DESC_STATIC(langid),
+};
+
+static status_t usb_register_cb(
+    void *cookie,
+    usb_callback_op_t op,
+    const union usb_callback_args *args
+)
+{
+    if (op == USB_CB_ONLINE) {
+        usbc_setup_endpoint(1, USB_IN, 0x40, USB_BULK);
+        usbc_setup_endpoint(1, USB_OUT, 0x40, USB_BULK);
+        usbc_setup_endpoint(2, USB_IN, 0x40, USB_INTR);
+
+        usb_online = true;
+    }
+    return NO_ERROR;
+}
+
+static status_t usb_setup_cb(
+    void *cookie,
+    usb_callback_op_t op,
+    const union usb_callback_args *args
+)
+{
+    if (op != USB_CB_SETUP_MSG) {
+        return ERR_NOT_IMPLEMENTED;
+    }
+
+    static uint8_t buf[EP0_MTU];
+    const struct usb_setup *setup = args->setup;
+    // NB: The host might send us some modem commands here. Since we're not a
+    // real modem, we're probably okay to drop them on the floor and reply with
+    // an acknowledgement.
+    if (setup->length) {  // Has data phase?
+        if (setup->request_type & 0x80) {
+            // We have to send data?
+            usbc_ep0_send(buf, setup->length, 64);
+        } else {
+            usbc_ep0_recv(buf, setup->length, NULL);
+            usbc_ep0_ack();
+        }
+    } else {
+        switch (setup->request) {
+            case CDC_REQ_SEND_CMD:
+            case CDC_REQ_GET_RESP:
+            case CDC_REQ_SET_COMM_FEATURE:
+            case CDC_REQ_GET_COMM_FEATURE:
+            case CDC_REQ_CLEAR_COMM_FEATURE:
+            case CDC_REQ_SET_LINE_CODING:
+            case CDC_REQ_GET_LINE_CODING:
+            case CDC_REQ_SET_CONTROL_LINE_STATE:
+            case CDC_REQ_SEND_BREAK:
+                // Ack any command that we understand.
+                usbc_ep0_ack();
+                break;
+        }
+    }
+
+    return NO_ERROR;
+}
+
+status_t cdcserial_start(void)
+{
+    usb_setup(&config, &usb_setup_cb);
```

Comment:
The target is really supposed to do the usb_setup, then call into <class>_start which adds their interfaces. This allows multiple classes to be present at a time.

That'll probably mess up your usb_setup_cb, since each class may register their own. Will probably need a list of callbacks, or feed them through the regular usb callback that was already there.